src/main/java/com/redhat/rhjmc/containerjfr/commands/internal/PortScanCommand.java

```diff
@@ -48,18 +51,44 @@ public boolean validate(String[] args) {
 
     @Override
     public void execute(String[] args) throws Exception {
-        scan().forEach(m -> cw.println(String.format("%s -> %s", m.hostname, m.ip)));
+        findCompatibleJvms().forEach(m -> cw.println(String.format("%s -> %s", m.hostname, m.ip)));
     }
 
     @Override
     public Output<?> serializableExecute(String[] args) {
         try {
-            return new ListOutput<>(scan());
+            return new ListOutput<>(findCompatibleJvms());
         } catch (Exception e) {
             return new ExceptionOutput(e);
         }
     }
 
+    private List<IpHostMapping> findCompatibleJvms() throws UnknownHostException, InterruptedException {
+        // Check for environment variables added by Kubernetes that indicate
+        // IP addresses listening on port 9091.
+        List<IpHostMapping> mapping = System.getenv().entrySet().parallelStream()
+                .filter(e -> e.getKey().endsWith(KUBERNETES_ENV_SUFFIX))
+                .map(e -> {
+                    try {
+                        return testHost(InetAddress.getByName(e.getValue()));
+                    } catch (IOException ignored) {
+                        return null;
+                    }
+                }).filter(m -> m != null).collect(Collectors.toList());
+        if (mapping.isEmpty()) {
+            // No matches from environment variables, use port scan
+            mapping = scan();
+        }
+        return mapping;
+    }
+
+    private IpHostMapping testHost(InetAddress addr) throws IOException {
+        try (Socket s = new Socket()) {
+            s.connect(new InetSocketAddress(addr, 9091), 100);
+            return new IpHostMapping(addr.getHostAddress(), addr.getCanonicalHostName());
```

Comment:
Does `getCanonicalHostName` here return a human-readable hostname (as it does in a plain Docker network), or do we still get back the cluster-internal IP?

If it's just the IP then I think a really nice enhancement would be to scan over the environment variables and extract the service names from them and use that to populate this structure.

Reply:
It is showing the cluser-internal hostnames for me now.
![Screenshot from 2019-06-21 17-56-07](https://user-images.githubusercontent.com/4326090/59953912-47898480-9450-11e9-8567-bea28897689d.png)